java/core/src/main/java/com/whylogs/core/metrics/BaseMetric.java

```diff
@@ -0,0 +1,13 @@
+package com.whylogs.core.metrics;
+
+import lombok.NonNull;
+
+public abstract class BaseMetric<TSubclass extends BaseMetric> extends Metric {
```

Comment:
Comment why we need this?

Reply:
This an application of the Curiously Recursive Template Pattern ([CRTP](https://nuah.livejournal.com/328187.html?) - which collections in Java also follow this) which allows us to have f-bound polymorphism so it's down at compile time not runtime. This does give a performance gain, but my practical reason for doing this is for useability. 


- It allows for method chaining like 
  m.zero().setMax(6) 
  
  Now that's a bit of a silly example, but without this the compiler wouldn't know that m is an IntsMetric and couldn't perform this without casting. You'd have to do a check and then go. 

- In the resolver will will have a list of Metrics this allows us to not have type eraser and not depend on passing strings around
- Well just more in general a lot of the passing of type names including custom types will be handled automatically with this as it allows us to use the typing system. The strings work great for python where its not strongly typed, but I recommend us utilizing the typing in Java to our advantage

It may be as we go that this isn't as useful and need it iterate, but I strongly think it worth exploring as we go through the registry and resolver stages to make full use of the language.

Reply:
^ oh also example the merge loses it's type to Metric if it's abstracted under the Metric class. Having it under the (currently named) BaseMetric is what allows us to do the merge while returning that type. 

As for the change in name - What about just AbstractMetric?